clar2wasm/tests/wasm-generation/principal.rs

```diff
@@ -0,0 +1,187 @@
+                        .expect("slice bigger than 20 bytes"),
+                ),
+                contract_name.into(),
+            )))
+        } else {
+            // standard principal requested
+            Value::Principal(PrincipalData::Standard(StandardPrincipalData(
+                version,
+                principal
+                    .as_slice()
+                    .try_into()
+                    .expect("slice bigger than 20 bytes"),
+            )))
+        }
+    }
+
+    fn create_error_construct(error_code: u8, principal_data: Option<Value>) -> Value {
+        Value::error(
+            TupleData::from_data(vec![
+                ("error_code".into(), Value::UInt(error_code.into())),
+                (
+                    "value".into(),
+                    Value::Optional(OptionalData {
+                        data: principal_data.map(Box::new),
+                    }),
+                ),
+            ])
+            .unwrap()
+            .into(),
+        )
+        .unwrap()
+    }
+
+    fn create_error_destruct(hash_bytes: Value, version_byte: u8) -> Value {
+        Value::error(
+            TupleData::from_data(vec![
+                ("hash-bytes".into(), hash_bytes),
+                ("name".into(), Value::Optional(OptionalData { data: None })),
+                (
+                    "version".into(),
+                    Value::Sequence(SequenceData::Buffer(BuffData {
+                        data: vec![version_byte],
+                    })),
+                ),
+            ])
+            .unwrap()
+            .into(),
+        )
+        .unwrap()
+    }
+
+    proptest! {
+        #![proptest_config(runtime_config())]
+
+        #[test]
+        fn crosscheck_principal_construct(
+            version_byte in 0x00..=0x1f,
+            hash_bytes in buffer(20)
+        ) {
+            let expected_principal = create_principal(
+                version_byte as u8,
+                &hash_bytes.clone().expect_buff(20).unwrap(),
+                None
+            );
+
+            let expected = match version_byte {
+                // Valid range for version_bytes
+                0x00..=0x1f => {
+                    match version_byte {
+                        // Since tests runs on a Testnet version,
+                        // version_bytes single_sig (0x1A) || multi_sig (0x15), for Testnet,
+                        // will return an Ok value.
+                        0x1A | 0x15 => Value::okay(expected_principal),
+                        _ => Ok(create_error_construct(0, Some(expected_principal))),
+                    }
+                },
+                _ => Ok(create_error_construct(1, None)),
+            }.unwrap();
+
+            crosscheck(
+                &format!("(principal-construct? 0x{:02X} {hash_bytes})", version_byte),
+                Ok(Some(expected)),
+            );
+        }
+    }
+
+    proptest! {
+        #![proptest_config(runtime_config())]
```

Comment:
qq: is there a reason to prefer having multiple `proptest!` vs putting all the tests in the same `proptest!` block?

Reply:
No, same output